docker/Dockerfile

```diff
@@ -0,0 +1,62 @@
+# See the file "COPYING" in the main distribution directory for copyright.
+
+# Layer to build Zeek.
+FROM debian:buster-slim AS build_zeek
+
+# Configure system for build.
+RUN apt-get -q update \
+ && apt-get install -q -y --no-install-recommends \
+     bind9 \
+     bison \
+     cmake \
+     flex \
+     g++ \
+     gcc \
+     libmaxminddb-dev \
+     libpcap-dev \
+     libssl-dev \
+     libz-dev \
+     make \
+     python3-minimal \
+     python3-dev \
+     swig \
+     ninja-build \
+     python3-pip \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+# Copy over the Zeek source tree.
+# NOTE: This assumes that we build in the context of the parent directory (the
+# Git checkout root). We e.g., ensure that in the `Makefile` in this directory.
+COPY . /src/zeek
```

Comment:
I was just tinkering with this Benjamin and I noticed a hiccup: if you use this setup on a local tree, particularly one in which you already have a build, then the fact that the tree gets copied into the image has two consequences:
- It makes for a pretty large build context shipped to dockerd — something like 2.5GB instead of 500MB.
- The existing build tree can trip up the Zeek build inside the image. I just got this:

```
Step 5/10 : RUN ./configure      --generator=Ninja      --build-type=Release  && ninja -C build install
 ---> Running in dd5251b613c0
Build Directory : build
...
-- Looking for kqueue - not found

********** Begin libkqueue External Project CMake Output ************

CMake Error: The current CMakeCache.txt directory /src/zeek/build/libkqueue-build/CMakeCache.txt is different than the directory /home/christian/devel/zeek/zeek/build/libkqueue-build where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "/src/zeek/auxil/libkqueue/CMakeLists.txt" does not match the source "/home/christian/devel/zeek/zeek/auxil/libkqueue/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.

*********** End libkqueue External Project CMake Output *************


CMake Error at cmake/FindKqueue.cmake:128 (message):
  libkqueue CMake configuration failed
Call Stack (most recent call first):
  CMakeLists.txt:488 (include)


-- Configuring incomplete, errors occurred!
See also "/src/zeek/build/CMakeFiles/CMakeOutput.log".
See also "/src/zeek/build/CMakeFiles/CMakeError.log".
Also, before re-running configure, consider cleaning the cache (removing the build directory) via `make distclean`
The command '/bin/sh -c ./configure      --generator=Ninja      --build-type=Release  && ninja -C build install' returned a non-zero code: 1
make: *** [Makefile:8: all] Error 1
```
So the current build tree confuses the new build. I'm thinking we should do a `make distclean` on the copied tree (ignoring failures) to increase the odds we're starting from a clean tree.
There's another thing we could do to minimize the build context, but it's a bit involved ... ideally we'd want the notion of a bind mount during the image build, mounting the Zeek source tree. But you cannot do bind mounts during the build, only when running the image. A workaround is to prepare an image for a Zeek build, then do the build in a container running that image, and then `docker commit` the result into a new image. This has the benefit that the source tree never ends up in the image or in any layers — but the overall approach becomes more complicated. Not sure it's worth the effort for 500MB vs 2.5GB.
We could table the size as a potential optimization for the future, but I do think we need to avoid having the build tree in the image.

Reply:
> So the current build tree confuses the new build. I'm thinking we should do a `make distclean` on the copied tree (ignoring failures) to increase the odds we're starting from a clean tree.

Yeah, copying over changes from the current tree can lead to that. I personally have a local `.dockerignore` file which contains e.g., `build` (which cuts down on the time to copy the context), but this file is hard to generalize since we don't know what people have in their trees. I will add a `make distclean` after the `COPY` to get rid of e.g., `build` which we would also be using to build to avoid the clash, but leaving everything else untouched.

> There's another thing we could do to minimize the build context, but it's a bit involved ... ideally we'd want the notion of a bind mount during the image build, mounting the Zeek source tree. But you cannot do bind mounts during the build, only when running the image. A workaround is to prepare an image for a Zeek build, then do the build in a container running that image, and then `docker commit` the result into a new image. This has the benefit that the source tree never ends up in the image or in any layers — but the overall approach becomes more complicated. Not sure it's worth the effort for 500MB vs 2.5GB.
> ...
> We could table the size as a potential optimization for the future, but I do think we need to avoid having the build tree in the image.

The workaround you describe sounds like a dedicated build stage 🙂. We already use a dedicated stage to build (`build_zeek`) and only copy over the Zeek installation tree in `/usr/local/zeek` into our `final` stage which will become the published image, so the source tree with whatever uncommitted changes should never end up in the final image.

Reply:
> I personally have a local .dockerignore file which contains e.g., build (which cuts down on the time to copy the context), but this file is hard to generalize since we don't know what people have in their trees.
That's a great point. How about we put in that `.dockerignore` for `build` anyway? I doubt it will ever collide with any other uses of a directory of that name, it reduce the build context significantly if one has it, and if folks have other build trees at the toplevel, the `distclean` still catches it — win/win?

Reply:
> That's a great point. How about we put in that .dockerignore for build anyway?

I could see s a single issue with that. While `.dockerignore` behaves pretty much like `.gitignore` there is no way to locally extend it (for Git there is also `.git/info/exclude`), so we'd always need to maintain a blessed listed of files or directories to ignore in the repo, or accept that this file might always contain unstaged changes while developing. That's not great if people e.g., keep multiple build trees or additional stuff around in their repos, so I personally would rather not add such a file to the repo. WDYT?

Reply:
I was thinking `build` is okay to include because our own setup makes that the default build tree. But it's just an optimization and imperfect when people do extra stuff in the tree, so feel free to skip it. We can always revisit.